core/iwasm/libraries/libc-builtin/libc_builtin_wrapper.c

```diff
@@ -252,113 +167,114 @@ _vprintf_wa(out_func_t out, void *ctx, const char *fmt, _va_list ap,
                     /* FIXME: do nothing for these modifiers */
                     goto still_might_format;
 
+                case 'o':
                 case 'd':
                 case 'i':
                 {
-                    int32 d;
+                    char buf[64];
+                    PREPARE_TEMP_FORMAT();
 
                     if (long_ctr < 2) {
+                        int32 d;
                         CHECK_VA_ARG(ap, int32);
                         d = _va_arg(ap, int32);
+                        n = snprintf(buf, sizeof(buf), fmt_buf, d);
                     }
                     else {
                         int64 lld;
+                        /* Make 8-byte aligned */
+                        ap = (_va_list)(((uintptr_t)ap + 7) & ~(uintptr_t)7);
                         CHECK_VA_ARG(ap, int64);
                         lld = _va_arg(ap, int64);
-                        if (lld > INT32_MAX || lld < INT32_MIN) {
-                            print_err(out, ctx);
-                            break;
-                        }
-                        d = (int32)lld;
+                        n = snprintf(buf, sizeof(buf), fmt_buf, lld);
                     }
 
-                    if (d < 0) {
-                        out((int)'-', ctx);
-                        d = -d;
-                        min_width--;
-                    }
-                    _printf_dec_uint(out, ctx, (uint32)d, padding, min_width);
+                    OUTPUT_TEMP_FORMAT();
                     break;
                 }
                 case 'u':
+                case 'p':
+                case 'x':
+                case 'X':
+                case 'c':
```

Comment:
The handling of these cases are similar to `case 'o'/'d'/'i'`, except `int32 d` to `uint32 u`, `int64 lld` to `uint64 llu`, seems that we can merge them together?